Resolve semantic inconsistencies for non traditional messaging (#1027)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/messaging.md

Although messaging systems are not as standardized as, e.g., HTTP, it is assumed that the following definitions are applicable to most of them that have similar concepts at all (names borrowed mostly from JMS):

A *message* is an envelope around a potentially empty payload.
This envelope may offer the possibility to convey additional metadata, often under the key/value form.
Messages can be delivered to 0, 1, or multiple consumers depending on the dispatching semantic of the protocol.

> Traditional messaging brokers, such as JMS, use the concept of topics when a message is dispatched to potentially multiple consumers and queues when a message is dispatched to a single consumer.

Reviewer: Kafka, for example, but it's not limited to Kafka, uses consumer groups. Each record (message in Kafka lingo) is sent to a single consumer per consumer group. So, it's both a topic and a queue depending on the consumer groups and their composition.

> Traditional messaging brokers, such as JMS, use the concept of topics when a message is dispatched to potentially multiple consumers and queues when a message is dispatched to a single consumer.

Reviewer 2: I see that this description was added below, thanks!

Although messaging systems are not as standardized as, e.g., HTTP, it is assumed that the following definitions are applicable to most of them that have similar concepts at all (names borrowed mostly from JMS):

A *message* is an envelope around a potentially empty payload.
This envelope may offer the possibility to convey additional metadata, often under the key/value form.
Messages can be delivered to 0, 1, or multiple consumers depending on the dispatching semantic of the protocol.
Traditional messaging brokers, such as JMS, use the concept of topics when a message is dispatched to potentially multiple consumers and queues when a message is dispatched to a single consumer.
In a messaging system such as Apache Kafka, consumer groups are used. Each record, or message, is sent to a single consumer per consumer group.

> Whether a specific message is processed as if it were sent to a topic or queue entirely depends on the consumer groups and their composition.

Reviewer 2: ```suggestion
Whether a specific message is processed as if it was sent to a topic or queue entirely depends on the consumer groups and their composition.
```

> Whether a specific record, or message, is processed as if it were sent to a topic or queue entirely depends on the consumer groups and their composition.

Reviewer 3: > Each record, or message
> Whether a specific record, or message

Just refer to them as messages, per semantics established in the previous sentences.

Author: Fixed, thanks

Although messaging systems are not as standardized as, e.g., HTTP, it is assumed that the following definitions are applicable to most of them that have similar concepts at all (names borrowed mostly from JMS):

A *message* usually consists of headers (or properties, or meta information) and an optional body. It is sent by a single message *producer* to:

> * Physically: some message *broker* (which can be e.g., a single server, or a cluster, or a local process reached via IPC). The broker handles the actual routing, delivery, re-delivery, persistence, etc. In some messaging systems the broker may be identical or co-located with (some) message consumers.

Author: For Kafka, there is a leader of the cluster, but all the instances are brokers. There is not a single broker.

> * Physically: some message *broker* (which can be e.g., a single server, or a cluster, or a local process reached via IPC). The broker handles the actual routing, delivery, re-delivery, persistence, etc. In some messaging systems the broker may be identical or co-located with (some) message consumers.

Reviewer 4: OK, but I don't see how this contradicts the wording here? Physically each single message is still sent to a single broker within the cluster.

> * Physically: some message *broker* (which can be e.g., a single server, or a cluster, or a local process reached via IPC). The broker handles the actual routing, delivery, re-delivery, persistence, etc. In some messaging systems the broker may be identical or co-located with (some) message consumers.

Author: I'm going to try and reword to re-add, as Kafka doesn't do routing in the brokers, it's done before it gets to the brokers

> * Physically: some message *broker* (which can be e.g., a single server, or a cluster, or a local process reached via IPC). The broker handles the actual routing, delivery, re-delivery, persistence, etc. In some messaging systems the broker may be identical or co-located with (some) message consumers.

Reviewer 4: The "routing" meant here is the routing of the received message to other brokers, the consumers, permanent storage, etc. The routing you mean is probably the client-side selection of a broker? But anyway, since it already says, "delivery, re-delivery, persistence, etc.", writing "routing" here is actually redundant/misleading, so I'm fine with removing that word.

> Traditional messaging brokers, such as JMS, use the concept of topics when a message is dispatched to potentially multiple consumers and queues when a message is dispatched to a single consumer.

Reviewer 4: It seems this should be merged/moved into the Destination section that follows.

Author: Will move and adjust

> In a messaging system such as Apache Kafka, consumer groups are used. Each record, or message, is sent to a single consumer per consumer group.

Author: I'll add "For instance, there can be multiple consumer groups retrieving records from the same topic in Kafka." to the end of the section.

Yes it acts as a queue if there is only one consumer group, but when there are multiple consumer groups it doesn't

> Whether a specific message is processed as if it was sent to a topic or queue entirely depends on the consumer groups and their composition.

Author: I think the extra line I'm adding, mentioned in the previous comment, will resolve the confusion

### Destinations

A destination is usually identified by some name unique within the messaging system instance, which might look like a URL or a simple one-word identifier.

> Traditional messaging involves two kinds of destinations: *topic*s and *queue*s.

Reviewer 4: Here I would expect a description of "consumer group" along with the "traditional" types.

With Apache Kafka, the physical broker a message is written to depends on the number of partitions, and which broker is the *leader* of the partition the record is written to.
* Logically: some particular message *destination*.

Messages can be delivered to 0, 1, or multiple consumers depending on the dispatching semantic of the protocol.

### Destinations

A destination is usually identified by some name unique within the messaging system instance, which might look like a URL or a simple one-word identifier.
Traditional messaging, such as JMS, involves two kinds of destinations: *topic*s and *queue*s.
A message that is sent (the send-operation is often called "*publish*" in this context) to a *topic* is broadcasted to all consumers that have *subscribed* to the topic.
A message submitted to a queue is processed by a message *consumer* (usually exactly once although some message systems support a more performant at-least-once mode for messages with [idempotent][] processing).

> In a messaging system such as Apache Kafka, all destinations are *topic*s.

Reviewer 4: Hmm, that wording is dangerous, you make readers think that they should use "destination_kind=topic"

Author: For Kafka, `destination_kind` should always be `topic`

First, the lower-level receiving of a message at a consumer, and then the logical processing of the message.
Often, the waiting for a message is not particularly interesting and hidden away in a framework that only invokes some handler function to process a message once one is received
(in the same way that the listening on a TCP port for an incoming HTTP message is not particularly interesting).

### Conversations

> In some messaging systems, a message can receive a reply message, or possibly multiple, that answers a particular other message that was sent earlier. All messages that are grouped together by such a reply-relationship are called a *conversation*.

Reviewer 4: ```suggestion
In some messaging systems, a message can receive one or more a reply messages that answers a particular other message that was sent earlier. All messages that are grouped together by such a reply-relationship are called a *conversation*.
```

```

The destination name SHOULD only be used for the span name if it is known to be of low cardinality (cf. [general span name guidelines](../api.md#span)).
This should not be statically derived from application code, as applications can use logical names internally that don't match the real destination name.

> Destination name can be statically derived from configuration in this case, as it should contain the mapping from logical internal names to the real destination name.

Reviewer 3: In the trivial cases the application code can serve as configuration. That is different from aliasing of what's in the config to what application calls the resource. 

> This should not be statically derived from application code

There can be situations where the logical name known to the application is the only thing that is accessible to the instrumentation. I don't think it's is wrong to the application to describe the span using the logical name of the destination (the physical identity, if available, can be added as attributes).

> Destination name can be statically derived from configuration in this case, as it should contain the mapping from logical internal names to the real destination name.

Author: If the span is described with a name known only to the application internally, what is the benefit of having that on a trace?

Wouldn't it require knowledge from the person interrogating the trace to understand it's an application internal name, and thus have some way of translating that into an actual name?

> Destination name can be statically derived from configuration in this case, as it should contain the mapping from logical internal names to the real destination name.

Reviewer 4: @kenfinnigan

> If the span is described with a name known only to the application internally, what is the benefit of having that on a trace?

The main purpose of the span name is to be able to group similar traces or spans together using an equal span name as the grouping criterion. So the span name should approximately indicate which code is run for that span.

Also, if you found a span to be important, you might send a trace to the development team for investigation, and they will be able to make sense of this "internal" name.

> Destination name can be statically derived from configuration in this case, as it should contain the mapping from logical internal names to the real destination name.

Author: I've revised the description to prefer real destination names over virtual/logical ones

The destination name SHOULD only be used for the span name if it is known to be of low cardinality (cf. [general span name guidelines](../api.md#span)).
This can be assumed if it is statically derived from application code or configuration.

> Wherever possible, the preference is to use real destination names over logical or aliased names.

Reviewer 4: What does "real" mean here? Would "canonical" or "resolved" be better?

Author: by "real" I meant the actual name of the topic/queue that the message is being sent to.

I'm not sure "canonical" or "resolved" align with "real". Maybe "actual"?

Reviewer 4: Maybe this would be more clear:

```suggestion
Wherever possible, the preference is to use real destination names after resolving logical or aliased names.
```

| -------------- | ----------- |
| `send`         | A message is sent to a destination by a message producer/client.       |
| `receive`      | A message is received from a destination by a message consumer/server. |

> | `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. |

Author: In that case I'd recommend changing the name to "processed".

For me the name "process" meant it was ready to process, not that it already had been

> | `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. |

Reviewer 4: > For me the name "process" meant it was ready to process, not that it already had been

`process` is simply what the span should set that traces the operation of the message being processed. So when you start the span, the message was ready to be processed from some unknown time before and by ending the span you signal that processing of the message is finished.

> | `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. |

Author: so the span is intended to go from "receipt" by the processor, to the point of "success" or completion?

If `process` is the preferred name, I can adjust the verbiage to make it clearer about what the operation name actually refers to

> | `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. |

Reviewer 4: > "from "receipt" by the processor,"

Well, I would not use "receipt" here, I would rather say "from the point where the processor gets passed the message". I think the usual case is that the processor is invoked with a previously received and (pre)parsed message object.

> | `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. |

Author: I *think* I've reverted this to the original definition

| `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. |

## Messaging attributes

| Attribute name |                          Notes and examples                            | Required? |
| -------------- | ---------------------------------------------------------------------- | --------- |
| `messaging.system` | A string identifying the messaging system such as `kafka`, `rabbitmq` or `activemq`. | Yes |
| `messaging.destination` | The message destination name, e.g. `MyQueue` or `MyTopic`. This might be equal to the span name but is required nevertheless. | Yes |
| `messaging.destination_kind` | The kind of message destination: Either `queue` or `topic`. | Yes, if either of them applies. |

> | `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). | If temporary (assumed to be `false` if missing). |

Author: "temp_destination" is meaningless in the sense that you may not know when performing the trace that a topic/queue is temporary or not.

That can happen internally within messaging systems. It's not possible to guarantee you know it is or isn't a temp destination.

In addition, does it really matter if it is or isn't temporary?

> | `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). | If temporary (assumed to be `false` if missing). |

Reviewer: Kafka does not do temporary topics. It can be pretty expensive to do so.

> | `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). | If temporary (assumed to be `false` if missing). |

Reviewer 4: > It's not possible to guarantee you know it is or isn't a temp destination.

> Kafka does not do temporary topics. It can be pretty expensive to do so.

OK to both statements, but why is this an argument for removing this optional attribute? It is still useful for traditional messaging systems at least.

> | `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). | If temporary (assumed to be `false` if missing). |

Author: Mostly because it's a security concern to expose that information.

If the consensus is it's not, I can add it back as optional

> | `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). | If temporary (assumed to be `false` if missing). |

Author: I was getting confused, it's not related to security, ignore me

> | `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

Author: But a Kafka Topic is very different from a JMS Topic, for instance.

That was the reason for this addition.

If it's set to "Topic", the meaning differs depending on what the messaging system type is. It's clearer to have a separate kind.

> | `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

Author: Put another way, how do we preclude a messaging system of "kafka" and a destination kind of "queue", as it's an invalid combination. The same with MQTT

> | `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

Reviewer 4: > how do we preclude a messaging system of "kafka" and a destination kind of "queue", as it's an invalid combination

We don't. How do we preclude a messaging system of "rabbitmq" and a destination kind of "kafka-topic"? Ultimately, the onus of using the semantic conventions correctly and producing semantically valid data is on the instrumentation writer.

> | `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

Reviewer 2: `queue` would just not be used in these cases but from the combination of system=`kafka`/`mqtt` and kind=`topic`, consumers aware of this distinction will be able to detect it properly. Would these two additions be the only ones you can think of or would you expect more to come here?

Apart from that, you made it required but kept it limited to the set of values defined, which will block uses where neither of them applies. If we extend the "enum" beyond topic and queue, I'd be fine with making it required but allow custom values here. This should be well-considered, however, since custom values usually don't have any semantics known to consumers and will therefore not be available for any automated analysis.

> | `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

Author: My main concern was the wrong thing being done and setting "kafka" and "queue", for instance.

If there's no way to prevent that, and there's a preference for not constraining the possible values for `destination_kind` I can remove the change

> | `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

Author: I've removed the change.

| `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |
| `messaging.protocol` | The name of the transport protocol such as `AMQP` or `MQTT`. | No |
| `messaging.protocol_version` | The version of the transport protocol such as `0.9.1`. | No |

> | `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. | No |

Author: The previous attribute was removed because it's a security hole, and is also often meaningless or undeterminable.

When sending a message to a broker that's clustered, how do you know exactly which URL the message went to?

> | `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. | No |

Reviewer: It's also not compatible with Kafka where you have bootstrap servers (plural). You don't know to which one you are actually sending (you write to the one hosting the leader of the partition of the topic on which you are writing).

> | `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. | No |

Author: I can add it back with a note it only applies to traditional messaging

| -------------- | ----------- |
| `send`         | A message is sent to a destination by a message producer/client.       |
| `receive`      | A message is received from a destination by a message consumer/server. |
| `process`      | A message that was previously received from a destination is received for processing by a message consumer/server. It spans receipt of the message to completion of processing, whether successful or not.|

## Messaging attributes

| Attribute name |                          Notes and examples                            | Required? |
| -------------- | ---------------------------------------------------------------------- | --------- |
| `messaging.system` | A string identifying the messaging system such as `kafka`, `rabbitmq` or `activemq`. | Yes |
| `messaging.destination` | The message destination name, e.g. `MyQueue` or `MyTopic`. This might be equal to the span name but is required nevertheless. | Yes |
| `messaging.destination_kind` | The kind of message destination: One of `queue`, `topic`, `kafka-topic`, or `mqtt-topic`. | Yes |

> | `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). Only applicable to traditional messaging systems. | If temporary (assumed to be `false` if missing). |

Reviewer 2: ```suggestion
| `messaging.temp_destination` | A boolean that is `true` if the message destination is [temporary](#temporary-destinations). | If temporary (assumed to be `false` if missing). |
```

I wouldn't say that, especially if the distinction is unclear. If a particular system has a concept of temporary destination, set the value, otherwise just don't - that's the only distinction I would make and which we already do.

| `messaging.protocol` | The name of the transport protocol such as `AMQP` or `MQTT`. | No |
| `messaging.protocol_version` | The version of the transport protocol such as `0.9.1`. | No |

> | `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. Only applicable to traditional messaging systems. | No |

Reviewer 2: ```suggestion
| `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. | No |
```
I wouldn't say that either here since there's no clear distinction. It's an optional attribute so it's fine to only set it if it fits and is reasonable/noteworthy.

Author: Can you clarify what you mean by "no clear distinction"?

Reviewer 2: There is no clear definition about what's considered to be a traditional or untraditional messaging system and even if there was, I don't think whether a connection string is applicable or not is dependent on that. Since it's an optional attribute, one would only set it if suitable and applicable anyway. If you think that for Kafka in particular it should be advised to _not_ set this, you could mention this in the Kafka section below including an explanation why this is the case.

Author: That's fine, I think I've removed the addition of the comment in the latest version

> | `messaging.service` | Name of the external broker, or name of the service being interacted with. | No |

Reviewer 2: ```suggestion
| `messaging.service` | Name of the external broker, or name of the service being interacted with. See note below for a definition. | No |
```

| `messaging.protocol_version` | The version of the transport protocol such as `0.9.1`. | No |
| `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. | No |
| `messaging.url` | Connection string such as `tibjmsnaming://localhost:7222` or `https://queue.amazonaws.com/80398EXAMPLE/MyQueue`. Only applicable to traditional messaging systems. | No |
| `messaging.service` | Name of the external broker, or name of the service being interacted with. | No |
| `messaging.message_id` | A value used by the messaging system as an identifier for the message, represented as a string. | No |
| `messaging.conversation_id` | The [conversation ID](#conversations) identifying the conversation to which the message belongs, represented as a string. Sometimes called "Correlation ID". | No |
| `messaging.message_payload_size_bytes` | The (uncompressed) size of the message payload in bytes. Also use this attribute if it is unknown whether the compressed or uncompressed payload size is reported. | No |
| `messaging.message_payload_compressed_size_bytes` | The compressed size of the message payload in bytes. | No |

> Additionally at least one of `net.peer.name` or `net.peer.ip` from the [network attributes][] is required and `net.peer.port` is recommended.

Reviewer 2: Did you remove this sentence on purpose? Why?

Author: `net.peer.name` or `net.peer.ip` are not required for Kafka, MQTT, etc.

I can change it to say it's recommended instead?

Reviewer 2: Wouldn't you just specify the hostname (or IP) of the broker that particular message is sent to here in case of Kafka? Or are you saying there are client libraries where this information is not available and therefore  instrumentations would merely be able to specify (one of) the bootstrap server(s) here?

Author: Such information as to where the message was sent to or received from, is not available in the Kafka clients.

As mentioned in another comment, I don't think bootstrap servers is the right information to be capturing instead

Reviewer 5: This is not technically correct. There is `org.apache.kafka.clients.consumer.Consumer#partitionsFor` which gives you a leader node for a given topic and partition. `Producer` gives the same metadata. `ConsumerRebalanceListener` can be used to refresh that metadata when it changes.

Thus you can get that information for Kafka. Is it feasible or not is totally separate question.

Reviewer: Yes, you can have an idea of the leader, however, this data may not be fresh as you may be checking while rebalancing is in progress (I'm not sure what would be the returned value in this case especially if the partition gets revoked). In general, in Kafka, don't need to know who's the leader of a partition.

Author: @iNikem Saying information is available somewhere is meaningless if it can't be proven to be accessible. From an instrumentation perspective, it might be possible to retrieve that information, but if it can't be done by frameworks or user applications, for me that makes it unusable.

Separate to that is the meaning. If it's an ephemeral value that is subject to frequent change, what is the value in capturing that information?

Why is there a concern about messaging systems such as Apache Kafka not defining these values?

Reviewer 5: `net.*` attributes, which by semantic convention can be applied to all spans, allow for constructing an accurate network-level map of monitored distributed system. Together with metrics it can help end-user to discover that these specific nodes has higher than usual error rate. And potentially even take a remedial action, such as take faulty nodes down.

>it's an ephemeral value that is subject to frequent change

Frequency of change is very subjective. You may have a fixed cluster of Kafka brokers. In fact, Kafka being a "storage service", I would expect Kafka nodes to live at least days, if not weeks, but certainly not minutes or hours. To call them "ephemeral" is an exaggeration.

Author: It might be a stretch, it's not a complete exaggeration.

What diagnosis can happen in a situation where a network-level map contains `net.*` attributes for nodes that are no longer present in the system?

I can certainly see there would be situations where it's necessary to "spin up" another Kafka broker to handle a spike in messages and then disappear again. Isn't that what Kubernetes offers, the power and ease to scale up and then down again?

Granted, there will definitely be peers that are essentially "always there", or near enough, but I still think that information should be a "recommendation" and not a "requirement"

Reviewer 5: I am fine with recommendation.

I hope you don't tear down your Kafka nodes after incoming spike ends. At least wait for all messages to be processed :) Also, event-sourcing systems cannot allow for node disappearing. You have to persist all relevant data. All in all, short-lived Kafka nodes in production is a novel concept to me.

Reviewer 4: > What diagnosis can happen in a situation where a network-level map contains net.* attributes for nodes that are no longer present in the system?

You can cross-reference logs, where the kafka nodes may log their IP when starting.

> `messaging.service` refers to the logical name of the external broker or messaging system where a message was sent to, or received from. In an environment such as Kubernetes, it would be the Kubernetes Service Name.

Reviewer 2: Can you please clarify how this differs from the `peer.service` [general span attribute](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes)?

Author: My impression of `peer.service` is a service running on a specific peer, or host, of a network.

In an environment like Kubernetes a single "peer", or node, can have many different services exposed on it. Meaning that sending a message to a Kafka service named "my-kafka" means more than a `peer.service` of "mycluster.kube", or some permutation thereof.

It's possible I'm missing something, but that's my impression, which is why I felt a separate `message.service` was needed. Granted there might be situations where they are the same, but in Kubernetes that would be rare

Reviewer 2: I would interpret `peer.service` to be the particular service this span is directed at, which also works if a given peer or host exposes multiple different services.

@anuraaga do you think we could clarify this further in the spec?

Author: @arminru am I correct in saying your interpretation of `peer.service` is actually the name of the service?

From my perspective, all the other `peer.*` or `net.peer.*` information relates to specific host or network node information, which is why I see `peer.service` as referring to a service on a specific host, as opposed to a virtual service name that could be clustered across many hosts/nodes

Reviewer 2: Yes. `net.*` is about lower-level networking information where `net.peer.*` is about the host on the other end. `peer.service` on the other hand is about the higher-level service. I don't think there's any other `peer.*` attribute.

Author: That's where I don't see it aligning.

A virtual service, such as in Kubernetes, happens to be running a node (peer) of the network but that doesn't define the name it's known by. Nodes (peers) come and go, but the name of the virtual service will be the same. It's in that context that I don't see `peer.service` being the same as `messaging.service`

Reviewer 5: `peer.service`
> SHOULD be equal to the actual service.name resource attribute of the remote service if any

Resource's `service.name`:
> Logical name of the service. MUST be the same for all instances of horizontally scaled services.

Therefore `peer.service` is not related anyhow to a physical or virtual node. Your definition of `messaging.service` duplicates the above definition of `peer.service`.

Reviewer 2: I agree with @iNikem. If you find the `peer.service` particularly useful for messaging service, you could add a note in this semantic convention referencing it and saying that it is recommended to be added for messaging services. There is no need, however, to duplicate it. This would only lead to confusion since instrumentation authors would not know if they should leave out `peer.service` in this case or provide the same value for both attributes, which wouldn't make sense either. Consumers will also likely pick up `peer.service` in a generic handler and would have to handle collisions if `messaging.service` having the same meaning would be set additionally.

Reviewer 4: I'm not sure about that. The peer.service attribute is something that is expected to be manually configured by the user, not automatically determined by the instrumentation.

See https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes 

>  Users can define what the name of a service is based on their particular semantics in their distributed system. Instrumentations SHOULD provide a way for users to configure this name.
...
> Examples of peer.service that users may specify:

CC @anuraaga who introduced peer.service in #652

Still it sounds like this concept is not specific to messaging, so maybe we should have peer.detected_service if peer.service is otherwise fitting.

Author: @iNikem did you mean `messaging.service` duplicates `peer.service` or Resource's `service.name`? Personally, it maybe overlaps with the latter and not the former.

I don't see `peer.service` being a great name for what it is, but if that's the preferred approach I will remove `messaging.service`

Reviewer 2: They actually belong together.
Given, for example, a client A sending a request to a server process B exposing a service called `X`:
A creates a span for the request and sets `peer.service="X"`
B, on the other end, sets `service.name="X"` on its resource, which will be added to a span it creates for tracking the processing of A's request

Author: Thanks @arminru, it hasn't been clear where Resource fits, that helps.

I will drop `message.service` and use `peer.service`

Reviewer 2: Thanks! Please add a note in the Kafka section saying it is recommended to set this and how to determine the appropriate value to make sure instrumentation writers are aware of it.

@anuraaga please verify if you think it's fine that instrumentations set this automatically (while still allowing users to override it, although I'm not really sure how this will work in practice).

`messaging.destination` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.
The routing key MUST be provided to the attribute `messaging.rabbitmq.routing_key`, unless it is empty.

#### Apache Kafka

> For Apache Kafka, the following messaging attributes are useful:

Reviewer: +1

| Attribute name |                          Notes and examples                            |
| -------------- | ---------------------------------------------------------------------- |

> | `messaging.kafka.message_key` | Differs from `messaging.message_id` in that it's not unique, and can be `null`. |

Reviewer: message keys are not ids. Kafka does not have message-id (well, not visible to the user at least, there are ids for de-duplication purposes).

> | `messaging.kafka.consumer_group` | Name of the Kafka Consumer Group that is handling the message. |

Reviewer: No, consumer only (for the consumer group).

Both producer and consumer could have a client.id.

> | `messaging.kafka.consumer_client_id` | Client Id for the Consumer within a Consumer Group that is handling the message. |

Reviewer: It would make sense to just use client_id, which would apply for both producer and consumer.

`messaging.destination` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.
The routing key MUST be provided to the attribute `messaging.rabbitmq.routing_key`, unless it is empty.

#### Apache Kafka

For Apache Kafka, the following messaging attributes are useful:

| Attribute name |                          Notes and examples                            |
| -------------- | ---------------------------------------------------------------------- |
| `messaging.kafka.message_key` | Differs from `messaging.message_id` in that it's not unique, and can be `null`. |
| `messaging.kafka.consumer_group` | Name of the Kafka Consumer Group that is handling the message. |
| `messaging.kafka.consumer_client_id` | Client Id for the Consumer within a Consumer Group that is handling the message. |
| `messaging.kafka.partition` | Partition the message is sent to. |

> 

Reviewer 2: ```suggestion
```

@kenfinnigan

`markdownlint` fails with:
```
./specification/trace/semantic_conventions/messaging.md:164 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

Author: Thanks @arminru!

`messaging.destination` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.
The routing key MUST be provided to the attribute `messaging.rabbitmq.routing_key`, unless it is empty.

#### Apache Kafka

For Apache Kafka, the following additional attributes are defined:

| Attribute name |                          Notes and examples                            |
| -------------- | ---------------------------------------------------------------------- |

> | `messaging.kafka.message_key` | Differs from `messaging.message_id` in that it's not unique, can be `null`, and doesn't have a pre-defined type. |

Reviewer 2: Which type does the attribute have? It has to be either of the attribute types defined here:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/common/common.md#attributes
We can't simply say "any" here because neither the API nor wire protocols support this.
You could define it as the string representation of whatever type the object has.

Author: Thanks

#### Apache Kafka

For Apache Kafka, the following additional attributes are defined:

<!-- semconv messaging.kafka -->
| Attribute  | Type | Description  | Example  | Required |
|---|---|---|---|---|
| `messaging.kafka.message_key` | string | Message keys in Kafka are used for grouping alike messages to ensure they're processed on the same partition. They differ from `messaging.message_id` in that they're not unique. If the key is `null`, the attribute MUST NOT be set. [1] | `myKey` | No |
| `messaging.kafka.consumer_group` | string | Name of the Kafka Consumer Group that is handling the message. Only applies to consumers, not producers. | `my-group` | No |
| `messaging.kafka.client_id` | string | Client Id for the Consumer or Producer that is handling the message. | `5` | No |
| `messaging.kafka.partition` | string | Partition the message is sent to. | `2` | No |

**[1]:** If the key type is not string, it's string representation has to be supplied for the attribute.
<!-- endsemconv -->

For Apache Kafka producers, [`peer.service`](./span-general.md#general-remote-service-attributes) SHOULD be set to the name of the broker or external service the message will be sent to.

> Consumers SHOULD identify the [Resource](../../resource/semantic_conventions/README.md#service) that received the message.

Reviewer 2: I don't understand this sentence. Please clarify.

Author: It was an attempt to indicate that the Consumer trace should have a `Resource` associated with it of the broker service that received the message.

If it's not important, I can remove, or alternative wording appreciated

Reviewer 2: With the wording in 2f442b3ff0, it now makes sense. Thanks!

> The `service.name` of a Consumer's Resource SHOULD match the `peer.service` of the Producer.

Reviewer 2: As per the sentence above, `peer.service` on the producer span refers to the broker the message is sent to. `service.name` on the consumer's resource will be its own service name and not the name of the broker.

Author: Ok, which is different that I think what was mentioned in another comment where it was stated the `peer.service` on producer should match `service.name` on Consumer.

Is there no direct link/association/matching needed? If not I can remove this bit

Reviewer 5: In this case, when there is an intermediary (broker), they indeed will not match.

`messaging.destination` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.
The routing key MUST be provided to the attribute `messaging.rabbitmq.routing_key`, unless it is empty.

#### Apache Kafka

For Apache Kafka, the following additional attributes are defined:

| Attribute name |                          Notes and examples                            |
| -------------- | ---------------------------------------------------------------------- |
| `messaging.kafka.message_key` | Differs from `messaging.message_id` in that it's not unique, and can be `null`. The type is a String representation of the type of the actual value. |
| `messaging.kafka.consumer_group` | Name of the Kafka Consumer Group that is handling the message. Only applies to consumers, not producers. |
| `messaging.kafka.client_id` | Client Id for the Consumer or Producer that is handling the message. |
| `messaging.kafka.partition` | Partition the message is sent to. |

> For Apache Kafka producers, `peer.service` should be set to the name of the broker or external service the message will be sent to.

Reviewer 2: ```suggestion
For Apache Kafka producers, [`peer.service`](./span-general.md#general-remote-service-attributes) SHOULD be set to the name of the broker or external service the message will be sent to.
```

This could also be achieved by using `- ref` with a note in the Kafka section of the YAML file.

> Consumers should set `service.name` as the Resource receiving the message. The value should match `peer.service` set by a Producer.

Reviewer 2: This sentence is unclear to me and I think it can be left out as it does not really add anything to the definition we have here:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/README.md#service

| Links |  |  |  |
| SpanKind | `PRODUCER` | `CONSUMER` | `CONSUMER` |
| Status | `Ok` | `Ok` | `Ok` |
| `net.peer.name` | `"ms"` | `"ms"` | `"ms"` |

> | `net.peer.port` | `1234` | `1234` | `1234` |

Author: Was deliberately removed as it's not reasonable to capture that information.

What is it meant to represent in a clustered broker situation? You don't know which of the possible brokers the message actually was processed by when sending a message

> | `net.peer.port` | `1234` | `1234` | `1234` |

Reviewer 2: This is to specify the endpoint to which the message was sent to (or received from) which should also be available in a clustered situation. Can you please elaborate?

> | `net.peer.port` | `1234` | `1234` | `1234` |

Author: If I'm sending/receiving messages with Kafka, there is no information on the consumer or producer side that knows what actual broker sent/received the message.

All the producer/consumer knows is it communicated with a particular bootstrap server to send/receive. That server is not the one that sent/received the message, nor is it constant as the cluster can be repartitioned.

> | `net.peer.port` | `1234` | `1234` | `1234` |

Reviewer 2: Would it still make sense to specify the bootstrap server as this is the direct counterpart for this span here, or would you not deem this information valuable?

> | `net.peer.port` | `1234` | `1234` | `1234` |

Author: I don't think so, as it could disappear at any point in time.

It's the initial point to which a consumer/producer retrieves the partitions, but where messages are sent can change as the partitioned cluster changes.

From a traceability perspective, I don't think the bootstrap server names would be meaningful

`messaging.destination` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.
The routing key MUST be provided to the attribute `messaging.rabbitmq.routing_key`, unless it is empty.

> #### Apache Kafka

Reviewer 6: We have a tombstone filled in the instrumentation currently, is it useful to have here?

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/f23ad29187ecea4e742e1dd6a5baeb84020298d4/instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaclients/KafkaProducerTracer.java#L52

If not it's ok and we can remove it from the Instrumentation. But this seems like the right time to get out instrumentation synced up with the spec.

Reviewer 6: I might also consider separating out a subfolder for messaging implementations as per #968 - there are so many of them that I worry a single doc just gets unwieldy.

Author: No problems with the delay, I appreciate everyone is super busy right now and context switching IS difficult.

I can see it being beneficial to identify if a particular trace is related to a tombstone record. I can add it to the Kafka section.

Can the splitting be done in a subsequent PR, or you'd prefer to include it here as well? I can certainly follow up with such a change.

Reviewer 6: Splitting in a separate PR sounds great, thanks

| Links |  |  |  |
| SpanKind | `PRODUCER` | `CONSUMER` | `CONSUMER` |
| Status | `Ok` | `Ok` | `Ok` |
| `net.peer.name` | `"ms"` | `"ms"` | `"ms"` |
| `net.peer.port` | `1234` | `1234` | `1234` |
| `messaging.system` | `"rabbitmq"` | `"rabbitmq"` | `"rabbitmq"` |
| `messaging.destination` | `"T"` | `"T"` | `"T"` |
| `messaging.destination_kind` | `"topic"` | `"topic"` | `"topic"` |
| `messaging.service` | `"myRabbitMQ"` | `"myRabbitMQ"` | `"myRabbitMQ"` |
| `messaging.operation` |  | `"process"` | `"process"` |

### Apache Kafka

Given is a process P, that publishes a message to a topic T1 on Apache Kafka.
One process, CA, receives the message and publishes a new message to a topic T2 that is then received and processed by CB. 

```
Process P:  | Span Prod1 |
--
Process CA:              | Span Rcv1 |
                                | Span Proc1 |
                                  | Span Prod2 |
--
Process CB:                           | Span Rcv2 |
```

| Field or Attribute | Span Prod1 | Span Rcv1 | Span Proc1 | Span Prod2 | Span Rcv2
|-|-|-|-|-|-|
| Span name | `"T1 send"` | `"T1 receive"` | `"T1 process"` | `"T2 send"` | `"T2 receive`" |

> | Parent |  | Span Prod1 | Span Prod1 |  | Span Prod2 |
> | Links |  |  | Span Rcv1 | Span Prod1 |  |

Reviewer 2: ```suggestion
| Parent |  | Span Prod1 | Span Rcv1 | Span Proc1 | Span Prod2 |
| Links |  |  | Span Prod1 |  |  |
```
Proc1's parent would be Rcv1 in this case as it's Proc1's direct predecessor and Prod1 would be added as a link.
Same for Prod2 being a child of Proc1.

Author: I can understand the reasoning for Proc1's parent being Rcv1, but Prod2 is not a child of Proc1.

The reasoning is that in Kafka the producing of a message is disconnected in time from when, or how often, that message is processed.

In this situation, Prod2 may be processed quickly, or it could be days/weeks/months before it's processed. From a span perspective, I don't think it should have a parent of Proc1 due to that timing disconnect.

Author: Yes it is performed as part of processing the first message. However, in reactive messaging and event-driven architectures there isn't always a "parent-child" connection, it's more like a correlation or link relationship. They *may* be "parent-child", but not always.

If it is added as a parent, and it hasn't had any further span, doesn't that leave the parent span/trace unclosed? Does that make sense for reporting?

Reviewer 4: Related #958 / CC @anuraaga wanted kinda the opposite there.

Author: Thanks @Oberon00, I knew there was a better term but couldn't think of it. Reading the issue made me realize it was "follows-from".

So for reactive messaging and event-driven, producing a new message while processing an existing message is more a "follows-from" relationship as opposed to parent-child, which is why I had used Links

| Status | `Ok` | `Ok` | `Ok` |
| `net.peer.name` | `"ms"` | `"ms"` | `"ms"` |
| `net.peer.port` | `1234` | `1234` | `1234` |
| `messaging.system` | `"rabbitmq"` | `"rabbitmq"` | `"rabbitmq"` |
| `messaging.destination` | `"T"` | `"T"` | `"T"` |
| `messaging.destination_kind` | `"topic"` | `"topic"` | `"topic"` |

> | `messaging.service` | `"myRabbitMQ"` | `"myRabbitMQ"` | `"myRabbitMQ"` |

Reviewer 2: ```suggestion
```

You've removed this attribute again in the meantime in favor of using `peer.service`. Same for the two examples below.

Author: Will remove, looks like I missed it in the examples

| `messaging.service` | `"myRabbitMQ"` | `"myRabbitMQ"` | `"myRabbitMQ"` |
| `messaging.operation` |  | `"process"` | `"process"` |

> | `messaging.message_id` | `"a1"` | `"a1"`| `"a1"` |

Reviewer 2: Please leave this as-is since Message IDs are fine here.

Author: Oops, will fix

| Links |  |  |  | Span Prod1 | Span Prod2 |
| SpanKind | `PRODUCER` | `PRODUCER` | `CONSUMER` | `CONSUMER` | `CONSUMER` |
| Status | `Ok` | `Ok` | `Ok` | `Ok` | `Ok` |
| `messaging.system` | `"kafka"` | `"kafka"` | `"kafka"` | `"kafka"` | `"kafka"` |
| `messaging.destination` | `"Q"` | `"Q"` | `"Q"` | `"Q"` | `"Q"` |

> | `messaging.destination_kind` | `"topic"` | `"topic"` | `"topic"` | `"topic"` | `"topic"` |

Reviewer 2: Ah, now I see why you were concerned about queue being used in conjunction with Kafka. These examples are rather intended to illustrate how the general concept of batch receiving and processing from a queue is intended to work rather than addressing technical details of a particular messaging system. Feel free to just use another value for `messaging.system` to fix this inconsistency of Kafka and queue being used in conjunction.
It would be great if you could add a separate example for Kafka in particular that shows appropriate values for attributes in the `messaging.kafka.*` namespace.
